### PR TITLE
Refactored ClonedPtr

### PR DESCRIPTION
**Pull request by hamishun**
- Documentation
- Modern c++14
- New utility function for cloning `clonePtr` if you want to clone without using ClonedPtr.
- New `makeClonedPtr` utility.
- Sync API with other std lib smart pointers
- Tests Written

### CODE REVIEW

File: include/helib/ClonedPtr.h [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+
+  /**
+   * @brief Copy construct by cloning the managed object based on the
+   * cloning policy.
+   * @param other The `ClonedPtr` object to be cloned.
+   **/
+  ClonedPtr(const ClonedPtr& other) : ptr(copy(other.ptr)) {}
+
+  /**
+   * @brief Move construct by taking onwership of the managed object by the
```
Comment (jlhcrawford) [inline include/helib/ClonedPtr.h:102]:
```suggestion
   * @brief Move construct by taking ownership of the managed object by the
```

File: include/helib/ClonedPtr.h [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+  {
+    if (this != &other) {
+      delete ptr;
+      ptr = copy(other.ptr);
+    }
+    return *this;
+  }
+
+  /**
+   * @brief Move construct by taking onwership of the managed object by the
```
Comment (jlhcrawford) [inline include/helib/ClonedPtr.h:124]:
```suggestion
   * @brief Move construct by taking ownership of the managed object by the
```

File: include/helib/ClonedPtr.h [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+  }
+
+  /**
+   * @brief convert `ClonedPtr` to `bool`, `false` if there is a managed object
+   * is else converts to `true`.
+   **/
+  explicit operator bool() const noexcept { return ptr != nullptr; }
+
+  /**
+   * @brief Dereference the smartpointer.
```
Comment (jlhcrawford) [inline include/helib/ClonedPtr.h:172]:
```suggestion
   * @brief Dereference the smart pointer.
```

File: include/helib/ClonedPtr.h [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+   * @param other `ClonedPtr` object to compare to.
+   * @return `true` if this `ClonedPtr` does not manage the same object as
+   * `other` else `false`.
+   **/
+  bool operator!=(const ClonedPtr& other) const { return ptr != other.ptr; }
+
+  /**
+   * @brief Less than another `ClonedPtr` object.
+   * @param other `ClonedPtr` object to compare to.
+   * @return `true` if the the address of the managed object is less than
```
Comment (jlhcrawford) [inline include/helib/ClonedPtr.h:226]:
```suggestion
   * @return `true` if the address of the managed object is less than
```

File: include/helib/ClonedPtr.h [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+   * that of the object managed by `other` else `false`.
```
Comment (jlhcrawford) [inline include/helib/ClonedPtr.h:227]:
```suggestion
   * the object managed by `other` else `false`.
```

File: include/helib/ClonedPtr.h [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+   **/
+  bool operator<(const ClonedPtr& other) const { return ptr < other.ptr; }
+
+  /**
+   * @brief Greater than another `ClonedPtr` object.
+   * @param other `ClonedPtr` object to compare to.
+   * @return `true` if the the address of the managed object is greater than
```
Comment (jlhcrawford) [inline include/helib/ClonedPtr.h:234]:
```suggestion
   * @return `true` if the address of the managed object is greater than that of
```

File: include/helib/ClonedPtr.h [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+   * that of the object managed by `other` else `false`.
```
Comment (jlhcrawford) [inline include/helib/ClonedPtr.h:235]:
```suggestion
   * the object managed by `other` else `false`.
```

File: include/helib/ClonedPtr.h [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+   **/
+  bool operator>(const ClonedPtr& other) const { return ptr > other.ptr; }
+
+  /**
+   * @brief Less than or equal to another `ClonedPtr` object.
+   * @param other `ClonedPtr` object to compare to.
+   * @return `true` if the the address of the managed object is less than
```
Comment (jlhcrawford) [inline include/helib/ClonedPtr.h:242]:
```suggestion
   * @return `true` if the address of the managed object is less than that of
```

File: include/helib/ClonedPtr.h [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+   * or equal to that of the object managed by `other` else `false`.
```
Comment (jlhcrawford) [inline include/helib/ClonedPtr.h:243]:
```suggestion
   * to that of the object managed by `other` else `false`.
```

File: include/helib/ClonedPtr.h [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+   **/
+  bool operator<=(const ClonedPtr& other) const { return ptr <= other.ptr; }
+
+  /**
+   * @brief Greater than or equal to another `ClonedPtr` object.
+   * @param other `ClonedPtr` object to compare to.
+   * @return `true` if the the address of the managed object is greater than
```
Comment (jlhcrawford) [inline include/helib/ClonedPtr.h:250]:
```suggestion
   * @return `true` if the address of the managed object is greater than
```

File: include/helib/ClonedPtr.h [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+   * or equal to that of the object managed by `other` else `false`.
```
Comment (jlhcrawford) [inline include/helib/ClonedPtr.h:251]:
```suggestion
   * to that of the object managed by `other` else `false`.
```

**Review by jlhcrawford — APPROVED** [review]
Looks good to me.

File: include/helib/ClonedPtr.h [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+ * @brief A smart pointer that `clones` the object it holds when it is copied.
+ * @tparam X The type of object managed by the pointer.
+ * @tparam Cloner A policy of how to clone the object pointed to, either `deep`
+ * or `shallow`. The default is `deep`.
+ **/
+template <typename X, typename Cloner = DeepCopy<X>>
+class ClonedPtr
+{
+public:
+  using element_type = X;
```
Comment (hamishun) [inline include/helib/ClonedPtr.h:77]: Need to write doc.

Comment (hamishun) [inline include/helib/ClonedPtr.h:77]:
```suggestion
public:
  /**
   * @brief The type of the managed object.
   **/
  using value_type = X;
```

File: include/helib/ClonedPtr.h [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+
+#include <utility> // std::swap
+/**
+ * @file ClonedPtr.h
+ * @brief Implementation of smart pointer with "deep cloning" semantics.
+ * The API should be familiar to those that use the standard library smart
+ * pointers (unique_ptr and shared_ptr).
+ * @note To work, a class needs a "clone" method that creates a new copy.
+ * The "const"-ness applies to both the pointer and the object pointed to
+ * as this kind of "deep copy" semantics, this makes the most sense.
```
Comment (esteffin) [inline include/helib/ClonedPtr.h:23]: Can you please add some documentation of the `CopiedPtr`?

Comment (hamishun) [reply include/helib/ClonedPtr.h]: Done.

File: include/helib/ClonedPtr.h [hunk 14]
```diff
@@ -0,0 +1,332 @@
...
+ * @file ClonedPtr.h
+ * @brief Implementation of smart pointer `ClonedPtr` with "cloning" semantics.
+ * The API should be familiar to those that use the standard library smart
+ * pointers (unique_ptr and shared_ptr).
+ *
+ * Two "cloning" policies are provided one for "deep cloning" (default) and the
+ * other "shallow cloning". For an object to be managed by `ClonedPtr`, a class
+ * needs a "clone" method for deep cloning that creates a raw pointer to a copy
+ * of the object or a copy constructor for shallow cloning.  An alias
+ * `CopiedPtr` is provided for convinience which is a `ClonedPtr` with shallow
```
Comment (hamishun) [inline include/helib/ClonedPtr.h:26]:
```suggestion
 * `CopiedPtr` is provided for convenience which is a `ClonedPtr` with shallow
```

File: include/helib/ClonedPtr.h [hunk 15]
```diff
@@ -0,0 +1,332 @@
...
+#define HELIB_CLONEDPTR_H
+
+#include <utility> // std::swap
+/**
+ * @file ClonedPtr.h
+ * @brief Implementation of smart pointer `ClonedPtr` with "cloning" semantics.
+ * The API should be familiar to those that use the standard library smart
+ * pointers (unique_ptr and shared_ptr).
+ *
+ * Two "cloning" policies are provided one for "deep cloning" (default) and the
```
Comment (hamishun) [inline include/helib/ClonedPtr.h:22]:
```suggestion
 * Two "cloning" policies are provided; one for "deep cloning" (default) and the
```

**Comment by faberga** [comment]
Jenkins run tests

**Pull request merged**